src/SimpleWMSBody.cpp

```diff
@@ -0,0 +1,600 @@
+      boost::posix_time::time_duration timeSinceStart;
+      boost::posix_time::ptime         startTime =
+          time - boost::posix_time::microseconds(time.time_of_day().fractional_seconds());
+      bool inInterval = utils::timeInIntervals(
+          startTime, mTimeIntervals, timeSinceStart, mIntervalDuration, mFormat);
+
+      if (mIntervalDuration != 0) {
+        startTime -= boost::posix_time::seconds(timeSinceStart.total_seconds() % mIntervalDuration);
+      }
+      std::string timeString = utils::timeToString(mFormat.c_str(), startTime);
+
+      // Select a WMS texture over the period of timeDuration if timespan is enabled.
+      if (mProperties->mEnableTimespan.get()) {
+        boost::posix_time::time_duration timeDuration =
+            boost::posix_time::seconds(mIntervalDuration);
+        boost::posix_time::ptime intervalAfter = getStartTime(startTime + timeDuration);
+        timeString += "/" + utils::timeToString(mFormat.c_str(), intervalAfter);
+      }
+
+      auto texture1 = mTextureFilesBuffer.find(timeString);
+      auto texture2 = mTexturesBuffer.find(timeString);
+      auto texture3 = mTextures.find(timeString);
+
+      // Only load textures those aren't stored yet.
+      if (texture1 == mTextureFilesBuffer.end() && texture2 == mTexturesBuffer.end() &&
+          texture3 == mTextures.end() && inInterval) {
+        // Load WMS texture to the disk.
+        mTextureFilesBuffer.insert(std::pair<std::string, std::future<std::string>>(timeString,
+            mTextureLoader.loadTextureAsync(timeString, mRequest, mActiveWMS.mLayers, mFormat)));
+      }
+    }
+
+    bool fileError = false;
+
+    // Check whether the WMS textures are loaded to the disk.
+    for (auto tex = mTextureFilesBuffer.begin(); tex != mTextureFilesBuffer.end(); ++tex) {
+      if (tex->second.wait_for(std::chrono::seconds(0)) == std::future_status::ready) {
+        std::string fileName = tex->second.get();
+
+        if (fileName != "Error") {
+          // Load WMS texture to memory
+          mTexturesBuffer.insert(std::pair<std::string, std::future<unsigned char*>>(
+              tex->first, mTextureLoader.loadTextureFromFileAsync(fileName)));
+        } else {
+          fileError = true;
+          // Load background texture instead.
+          mTexturesBuffer.insert(std::pair<std::string, std::future<unsigned char*>>(
+              tex->first, mTextureLoader.loadTextureFromFileAsync(mBackgroundTextureFile)));
+        }
+        mTextureFilesBuffer.erase(tex);
+      }
+    }
+
+    // Check whether the WMS textures are loaded to the memory.
+    for (auto tex = mTexturesBuffer.begin(); tex != mTexturesBuffer.end(); ++tex) {
+      if (tex->second.wait_for(std::chrono::seconds(0)) == std::future_status::ready) {
+        mTextures.insert(std::pair<std::string, unsigned char*>(tex->first, tex->second.get()));
+        mTexturesBuffer.erase(tex);
+      }
+    }
```

Comment:
Here's the reason for the crash: You cannot use a for-loop if you're erasing elements from the container you're iterating over.

```suggestion
    // Check whether the WMS textures are loaded to the disk.
    auto fileIt = mTextureFilesBuffer.begin();
    while (fileIt != mTextureFilesBuffer.end()) {
      if (fileIt->second.wait_for(std::chrono::seconds(0)) == std::future_status::ready) {
        std::string fileName = fileIt->second.get();

        if (fileName != "Error") {
          // Load WMS texture to memory
          mTexturesBuffer.insert(std::pair<std::string, std::future<unsigned char*>>(
              fileIt->first, mTextureLoader.loadTextureFromFileAsync(fileName)));
        } else {
          fileError = true;
          // Load background texture instead.
          mTexturesBuffer.insert(std::pair<std::string, std::future<unsigned char*>>(
              fileIt->first, mTextureLoader.loadTextureFromFileAsync(mBackgroundTextureFile)));
        }

        fileIt = mTextureFilesBuffer.erase(fileIt);
      } else {
        ++fileIt;
      }
    }

    // Check whether the WMS textures are loaded to the memory.
    auto texIt = mTexturesBuffer.begin();
    while (texIt != mTexturesBuffer.end()) {
      if (texIt->second.wait_for(std::chrono::seconds(0)) == std::future_status::ready) {
        mTextures.insert(std::pair<std::string, unsigned char*>(texIt->first, texIt->second.get()));
        texIt = mTexturesBuffer.erase(texIt);
      } else {
        ++texIt;
      }
    }
```